File: src/main/java/ltdjms/discord/aichat/services/AIChatService.java
package ltdjms.discord.aichat.services;

import java.util.List;

import ltdjms.discord.aiagent.domain.ConversationMessage;
import ltdjms.discord.shared.DomainError;
import ltdjms.discord.shared.Result;

/** AI 聊天服務介面，負責處理 AI 請求並產生回應內容。 */
public interface AIChatService {

  /**
   * 生成 AI 回應內容（可能分段）。
   *
   * @param guildId Discord 伺服器 ID
   * @param channelId Discord 頻道 ID
   * @param userId 使用者 ID
   * @param userMessage 使用者訊息
   * @return 分割後的回應訊息或錯誤
   */
  Result<List<String>, DomainError> generateResponse(
      long guildId, String channelId, String userId, String userMessage);

  /**
   * 生成流式 AI 回應內容。
   *
   * @param guildId Discord 伺服器 ID
   * @param channelId Discord 頻道 ID
   * @param userId 使用者 ID
   * @param userMessage 使用者訊息
   * @param handler 流式回應處理器
   */
  void generateStreamingResponse(
      long guildId,
      String channelId,
      String userId,
      String userMessage,
      StreamingResponseHandler handler);

  /**
   * 生成流式 AI 回應內容（帶訊息 ID）。
   *
   * @param guildId Discord 伺服器 ID
   * @param channelId Discord 頻道 ID
   * @param userId 使用者 ID
   * @param userMessage 使用者訊息
   * @param messageId 觸發訊息的 ID
   * @param handler 流式回應處理器
   */
  void generateStreamingResponse(
      long guildId,
      String channelId,
      String userId,
      String userMessage,
      long messageId,
      StreamingResponseHandler handler);

  /**
   * 使用對話歷史生成 AI 回應（用於多輪工具調用）。
   *
   * @param guildId Discord 伺服器 ID
   * @param channelId Discord 頻道 ID
   * @param userId 使用者 ID
   * @param history 對話歷史
   * @param handler 流式回應處理器
   */
  void generateWithHistory(
      long guildId,
      String channelId,
      String userId,
      List<ConversationMessage> history,
      StreamingResponseHandler handler);
}

File: packages/ai/src/services/ai-chat-service.ts
import { type Result, type DomainError } from '@ltdjms/shared';
import { AIServiceConfig } from '../config/ai-service-config.js';

// ===== Chunk Types =====

/**
 * Type of streaming chunk.
 */
export enum StreamChunkType {
  REASONING = 'REASONING',
  CONTENT = 'CONTENT',
  TOOL_INTENT = 'TOOL_INTENT',
}

/**
 * A single chunk from the AI streaming response.
 */
export interface StreamChunk {
  type: StreamChunkType;
  content: string;
}

// ===== AIChatService Interface =====

/**
 * Streaming response handler callback interface.
 * Matches Java StreamingResponseHandler.
 */
export interface StreamingResponseHandler {
  onChunk(chunk: string, isComplete: boolean, error: DomainError | null): void;
  onChunkWithType(
    chunk: string,
    isComplete: boolean,
    error: DomainError | null,
    type: StreamChunkType,
  ): void;
}

/**
 * AI Chat Service interface.
 * Matches Java AIChatService interface.
 */
export interface AIChatService {
  /** Configuration for this service. */
  readonly config: AIServiceConfig;

  /**
   * Generates a non-streaming response.
   * Returns the complete response as an array of message strings (for pagination).
   */
  generateResponse(
    guildId: string,
    channelId: string,
    userId: string,
    userMessage: string,
  ): Promise<Result<string[], DomainError>>;

  /**
   * Generates a streaming response.
   *
   * @param agentEnabled - When true, includes agent (tool-calling) prompts in context
   */
  generateStreamingResponse(
    guildId: string,
    channelId: string,
    userId: string,
    userMessage: string,
    handler: StreamingResponseHandler,
    agentEnabled?: boolean,
  ): Promise<void>;

  /**
   * Generates a streaming response with a message ID for editing.
   *
   * @param agentEnabled - When true, includes agent (tool-calling) prompts in context
   */
  generateStreamingResponseWithId(
    guildId: string,
    channelId: string,
    userId: string,
    userMessage: string,
    messageId: string,
    handler: StreamingResponseHandler,
    agentEnabled?: boolean,
  ): Promise<void>;

  /**
   * Generates a response with history (for thread-level conversations).
   */
  generateWithHistory(
    guildId: string,
    channelId: string,
    userId: string,
    history: Array<{ role: string; content: string }>,
    handler: StreamingResponseHandler,
  ): Promise<void>;
}

// ===== Allowed Channel Types =====

export interface AllowedChannel {
  guildId: string;
  channelId: string;
  channelName: string;
}

export interface AllowedCategory {
  guildId: string;
  categoryId: string;
  categoryName: string;
}

export interface AIChannelRestriction {
  channels: AllowedChannel[];
  categories: AllowedCategory[];
}

// ===== Routing Types =====

export enum Route {
  AGENT_ROUTE = 'AGENT_ROUTE',
  AI_CHAT_ROUTE = 'AI_CHAT_ROUTE',
  DENY = 'DENY',
}

export enum Source {
  AGENT_CONFIG = 'AGENT_CONFIG',
  CHANNEL_ALLOWLIST = 'CHANNEL_ALLOWLIST',
  CATEGORY_ALLOWLIST = 'CATEGORY_ALLOWLIST',
  AGENT_CONFIG_UNAVAILABLE = 'AGENT_CONFIG_UNAVAILABLE',
  NO_ALLOWLIST = 'NO_ALLOWLIST',
}

export interface Decision {
  route: Route;
  source: Source;
  detail?: string;
}

// ===== Message Splitter Types =====

// This is the raw message splitter limit (before markdown pipeline).
// DiscordMarkdownPaginator uses 1900 to leave room for pipeline processing (code fences, etc.).
export const MAX_MESSAGE_LENGTH = 1980;

// ===== Agent Config Types =====

export interface AIAgentChannelConfig {
  guildId: string;
  channelId: string;
  enabled: boolean;
  updatedAt: Date;
}

// ===== Conversation Memory Types =====

export interface ToolCallEntry {
  timestamp: Date;
  toolName: string;
  parameters: Record<string, unknown>;
  success: boolean;
  memorySummary: string;
  redactionMode: RedactionMode;
}

export enum RedactionMode {
  NONE = 'NONE',
  REDACTED = 'REDACTED',
  OMITTED = 'OMITTED',
}

export enum ConversationIdStrategy {
  THREAD_LEVEL = 'THREAD_LEVEL',
  MESSAGE_LEVEL = 'MESSAGE_LEVEL',
}

// ===== Tool Definition Types =====

export interface ToolDefinition {
  name: string;
  description: string;
  parameters: ToolParameter[];
}

export interface ToolParameter {
  name: string;
  type: 'STRING' | 'NUMBER' | 'ARRAY';
  description: string;
  required: boolean;
  defaultValue?: unknown;
}

export interface PermissionSetting {
  id: string;
  type: 'role' | 'member';
  allow?: bigint;
  deny?: bigint;
  allowSet?: string[];
  denySet?: string[];
}

export interface ModifyPermissionSetting {
  id: string;
  type: 'role' | 'member';
  allow?: bigint;
  deny?: bigint;
  allowSet?: string[];
  denySet?: string[];
}

export interface RoleCreateInfo {
  name: string;
  color?: string;
  permissions?: PermissionSetting[];
}

// ===== Tool Execution Context =====

export interface ToolExecutionContext {
  guildId: string;
  channelId: string;
  userId: string;
}

// ===== Domain Events for AI =====

export interface AgentConfigUpdatedEvent {
  guildId: string;
  channelId: string;
  enabled: boolean;
}

export interface AIMessagePublishedEvent {
  guildId: string;
  channelId: string;
  userId: string;
  userMessage: string;
  aiResponse: string;
  timestamp: Date;
}
